Enable health checks for Kubernetes with virtual defaults

rana opened the pull request:
Health checks are enabled for all Kubernetes clusters by default.

A design of creating one health check config default per resource is implemented. The choice eases adoption of health checks, supports existing clusters that already have database health checks, and avoids migrating the backend database. A new Kubernetes-specific `default-kube` health check config is added. And a database-specific `default` health check config already exists, and is preserved.

A virtual default design is implemented by returning health check configs from memory if they don't exist in the backend database. The approach has the benefit of not re-inserting default values to the backend after they're deleted, which a prior approach had.

Virtual defaults are added at the local health check service level, and returned from functions `GetHealthCheckConfig` and `ListHealthCheckConfigs`. Virtual defaults may be written, updated, and deleted to and from the backend. While virtual defaults may be deleted, it has the net effect of resetting the config to default settings, and matching all resources of that type (db, kube). Virtual defaults are always returned from health check `get` and `list` functions.

In this PR:
- Added `default-kube` health check config specific to Kubernetes only
- Updated local service functions `GetHealthCheckConfig` and `ListHealthCheckConfigs` to return virtual defaults
- Added unit tests
- Updated health check documentation with `default-kube` and info about virtual defaults

Part of #58413
Relates to:
- #58413

espadolini commented on lib/auth/healthcheckconfig/healthcheckconfigv1/service.go:
```diff
@@ -102,6 +109,15 @@ func (s *Service) GetHealthCheckConfig(ctx context.Context, req *healthcheckconf
+			// Return a new preset instance when needed.
+			switch req.GetName() {
+			case teleport.PresetDefaultHealthCheckConfigDBName:
+				return services.NewPresetHealthCheckConfigDB(), nil
```
I don't think we can touch the nature of `default` at this point.

espadolini reviewed (dismissed):
The `healthcheck.manager` relies on working events for `health_check_config`, so the approach we used for `installers` (where we _cannot_ generate events correctly because the virtual resource depends on the state of other resources, so events are broken and virtual resources are reified in the auth grpc layer) doesn't quite work; I've moved things around a bit and added events in #60552, PTAL.

rana commented on constants.go:
```diff
@@ -775,9 +775,15 @@ const (
+	// PresetDefaultHealthCheckConfigKubeName is the name of a preset
+	// health_check_config that enables health checks for all
+	// Kubernetes resources.
+	PresetDefaultHealthCheckConfigKubeName = "default_kube"
```
Okay, I'll switch it kebab-case naming with `default-kube`.

espadolini commented on lib/auth/init_test.go:
```diff
@@ -141,7 +140,7 @@ func TestBadIdentity(t *testing.T) {
 	// bad cert type
 	_, err = state.ReadSSHIdentityFromKeyPair(priv, pub)
-	require.IsType(t, trace.BadParameter(""), err)
+	require.ErrorIs(t, trace.BadParameter("failed to parse server certificate: not an SSH certificate"), err)
```
We don't care about capturing a trace here, and `errors.Is` takes the error first and then the comparison target.

```suggestion
	require.ErrorIs(t, err, &trace.BadParameterError{Message: "failed to parse server certificate: not an SSH certificate"})
```

That being said, this feels like a very specific check and we should probably relax it a bit like


```suggestion
	require.ErrorAs(t, err, new(*trace.BadParameterError))
	require.ErrorContains(t, err, "not an SSH certificate")
```

That being said, these changes seem a bit spurious and could be reverted here and addressed in a separate PR.

Commit:
Enable health checks for Kubernetes with virtual defaults

Health checks are enabled for all Kubernetes clusters by default.

A design of creating one health check config default per resource is implemented. The choice eases adoption of health checks, supports existing clusters that already have database health checks, and avoids migrating the backend database. A new Kubernetes-specific `default-kube` health check config is added. And a database-specific `default` health check config already exists, and is preserved.

A virtual default design is implemented by returning health check configs from memory if they don't exist in the backend database. The approach has the benefit of not re-inserting default values to the backend after they're deleted, which a prior approach had.

Virtual defaults are added at the local health check service level, and returned from functions `GetHealthCheckConfig` and `ListHealthCheckConfigs`. Virtual defaults may be written, updated, and deleted to and from the backend. While virtual defaults may be deleted, it has the net effect of resetting the config to default settings, and matching all resources of that type (db, kube). Virtual defaults are always returned from health check `get` and `list` functions.

Changes:
- Added `default-kube` health check config specific to Kubernetes only
- Updated local service functions `GetHealthCheckConfig` and `ListHealthCheckConfigs` to return virtual defaults
- Added unit tests
- Updated health check documentation with `default-kube` and info about virtual defaults

Part of #58413

Co-authored-by: Edoardo Spadolini <edoardo.spadolini@goteleport.com>

okraport reviewed (commented):
This looks good to me. Nice job. I would like @espadolini and @rosstimothy stamp before approving to be safe.

espadolini commented on lib/services/local/health_check_config_test.go:
```diff
@@ -211,5 +486,71 @@ func TestHealthCheckConfigParser(t *testing.T) {
+func cloneVirtualDefaultKube() *healthcheckconfigv1.HealthCheckConfig {
+	// Clone to avoid race conditions in parallel tests.
+	return proto.Clone(services.VirtualDefaultHealthCheckConfigKube()).(*healthcheckconfigv1.HealthCheckConfig)
+}
```
The `services.VirtualDefaultHealthCheckConfigFoo()` functions return a fresh copy every time, we don't need a `proto.Clone`.

rana commented on lib/cache/health_check_config_test.go:
```diff
@@ -57,13 +59,84 @@ func TestHealthCheckConfig(t *testing.T) {
 		},
 		deleteAll: p.healthCheckConfig.DeleteAllHealthCheckConfigs,
-		cacheList: p.cache.ListHealthCheckConfigs,
+		cacheList: filterHealthCfgNonVirtual(p.cache.ListHealthCheckConfigs),
```
The TODO is added.
